### PR TITLE
Fix testem runner, and further improvements

### LINKED ISSUE #3326: [$200] Introduce cross-browser CI testing
Travis CI is only running tests on Node.js. This is not sufficient for fast-paced cross-browser CI testing, and quickly ensuring changes are not breaking across all supported browsers.

I would like to suggest we get Fabric.js running on saucelabs open-source offering: https://saucelabs.com/open-source

Here's instructions for hooking into Travis:
https://docs.travis-ci.com/user/gui-and-headless-browsers/

I have managed to run Fabric.js tests using the testem test runner:
screen shot 2016-10-04 at 9 33 01 pm
As you can see, there are failing tests that need to be looked into.

Testem is a great tool for test-driven-development across multiple browsers at the same time.

We can set up a similar testem setup for hooking into saucelabs, like Ember.js:
https://github.com/emberjs/ember.js/blob/master/testem.dist.json

<svg xmlns="http://www.w3.org/2000/svg" width="168" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><mask id="a"><rect width="168" height="20" rx="3" fill="#fff"/></mask><g mask="url(#a)"><path fill="#555" d="M0 0h87v20H0z"/><path fill="#129e5e" d="M87 0h81v20H87z"/><path fill="url(#b)" d="M0 0h168v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11"><text x="43.5" y="15" fill="#010101" fill-opacity=".3">bountysource</text><text x="43.5" y="14">bountysource</text><text x="126.5" y="15" fill="#010101" fill-opacity=".3">$200 bounty</text><text x="126.5" y="14">$200 bounty</text></g></svg>

### PR DESCRIPTION
**Pull request by jedateach**
This is some work to get testem working again, in parallel with the existing approach.
It's a work-in-progress, but I wanted to start a pull request to share what I've been doing.

## Improvements in these changes:

- Fixes testem test runner
- CI testing on Chrome, in addition to Node (with Firefox to come close behind)
- Chrome, Firefox configured to run in headless mode
- `npm run testem` script for testem TDD
- `npm run testem:ci` script to hopefully replace `npm test`
- `npm run test:node` script to only run node tests
- Upgrades qunit to version 2.4.1 for testem, and switches to `node-qunit` for existing test script.
- Configured sensible qunit testTimeout (10 seconds) and noglobals=true configs
- Fixes one test that needed to be asynchronous

## TODO:

- [x] Ensure travis builds can use testem by updating Travis config
- [x] Resolve iText tests hanging
- [x] Resolve Firefox failures
- [ ] Decide how to do code coverage in this setup

---

Helps to resolve #3326 
Note however, that these changes do not make use of Saucelabs.

### CODE REVIEW

**Comment by jedateach** [comment]
Note there appears to be just one Firefox failure in travis:
```
not ok 847 Firefox 56.0 - fabric.Pattern: toObject
    ---
        actual: >
            function() {return fabric.document.createElement('canvas');}
        expected: >
            function () {return fabric.document.createElement('canvas');}
        stack: >
            @http://localhost:8080/9138/test/unit/pattern.js:86:5
            runTest@http://localhost:8080/node_modules/qunit/qunit/qunit.js:1478:16
            run@http://localhost:8080/node_modules/qunit/qunit/qunit.js:1464:6
            runTest/<@http://localhost:8080/node_modules/qunit/qunit/qunit.js:1670:7
            advance@http://localhost:8080/node_modules/qunit/qunit/qunit.js:1116:6
            begin@http://localhost:8080/node_modules/qunit/qunit/qunit.js:2937:4
            internalStart/config.timeout<@http://localhost:8080/node_modules/qunit/qunit/qunit.js:2051:6
            
        Log: |
    ...
```
...which seems easily fixable.


I get other failures when I run locally however - (7 in total).
One particular failure:

```
not ok 2252 Firefox 57.0 - fabric.util: get bounds of arc
    ---
        actual: >
            Max depth.,Max depth.,Max depth.,Max depth.
        expected: >
            Max depth.,Max depth.,Max depth.,Max depth.
        stack: >
            @http://localhost:7357/1345/test/unit/util.js:915:5
            runTest@http://localhost:7357/node_modules/qunit/qunit/qunit.js:1478:16
            run@http://localhost:7357/node_modules/qunit/qunit/qunit.js:1464:6
            runTest/<@http://localhost:7357/node_modules/qunit/qunit/qunit.js:1670:7
            advance@http://localhost:7357/node_modules/qunit/qunit/qunit.js:1116:6
            begin@http://localhost:7357/node_modules/qunit/qunit/qunit.js:2937:4
            internalStart/config.timeout<@http://localhost:7357/node_modules/qunit/qunit/qunit.js:2051:6

        message: >
            bounds are as expected
        Log: |
    ...
```

File: test.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-var testrunner = require('qunit');
+var testrunner = require('node-qunit');
```
Comment (asturur) [inline test.js]: so what is the big difference here? rewriting tests i was about to make the switch, but then i did not quite figure out what was the newer, and i just moved out from the old syntax of qunit 0.9.

**Comment by asturur** [comment]
I have a couple of questions:
I recently updated all the tests to run on a newer qunit, i quite did not understand on what version of qunit i was running. Why i was not using the node version, and why to switch from 1 to 2 you did not require any breaking changes?

Do you need help for the skipped tests? i can look into them.

Do you think is possible to run a second test page, or a different script to automate this page:
http://fabricjs.com/test/automated_visual/stroke.html ( that need to be extended with many other cases )

Saucelabs was offering older browsers support, this is a plus, is not a big deal.
Current fabricjs errors and bugs come from:
1 - changes in event handling and mouse interaction ( most common, harder to test )
2 - rendering optimization or `mostly harmless` changes that then reflect in visual bug ( less common, testable )

The firefox failures are probably due to different roundings.

**Comment by asturur** [comment]
in the old `npm run test` i was getting also some coverage information at which i was often looking into.
Can we make this other qunit spit it out?

**Comment by jedateach** [comment]
Ok, I've done some more research, and I think I can clarify.

The `qunit` package was before a node-based tool for running qunit tests. It relied on the `qunitjs` package, being the actual QUnit library. The `qunit` package relied on `qunitjs` version `2.1.1`

Screenshot from the last master travis build:
https://travis-ci.org/kangax/fabric.js/jobs/322020592
```
├─┬ qunit@1.0.0
│ ├── argsparser@0.0.7
│ ├─┬ cli-table@0.3.1
│ │ └── colors@1.0.3
│ ├── co@4.6.0
│ ├── istanbul@0.4.5 deduped
│ ├── qunitjs@2.1.1
│ ├── tracejs@0.1.8
│ └── underscore@1.8.3
```

The QUnit maintainers must have obtained ownership of the `qunit` package name, and started using it for the QUnit library itself from version `2.4.1`. That meant that the previous `qunit` package was moved to a new `node-qunit` package.

So yes - you were on version 2 of QUnit, included by the `qunit` package at v1.0.0.

Hopefully my changes help to clarify the split between the `qunit` library, and the `node-qunit` test runner.

**Comment by jedateach** [comment]
Regarding the skipped tests: yes please can you help resolve those. Can you un-skip these and get the tests hanging on Node?
(reversing commit e5ac487a030690c8f98d1263def73251a16bcada will do that for you)

Here's how you can isolate Testem to run Node only:
`npm run testem -- -l "Node"`
or
`npm run test:node`

**Comment by jedateach** [comment]
I've committed a reversal of those skipped tests, and you can see the Travis job hanging here:
https://travis-ci.org/kangax/fabric.js/jobs/323140803

**Comment by jedateach** [comment]
We should be able to get a visual regression test suite running in CI using chrome headless. Where is the code for the jasmine-based visual testing?

Here's a tool that could be easier to use than Jasmine: https://github.com/garris/BackstopJS

**Comment by jedateach** [comment]
With mouse/event issues - is there a library that abstracts these issues away?

Rendering optimizations - is that to do with the visual tests you mentioned above?

With Firefox rounding, perhaps we can make use of a qunit extension like https://github.com/JamesMGreene/qunit-assert-close ?

**Comment by jedateach** [comment]
Code coverage - I'm not sure about this yet.
https://github.com/gotwarlost/istanbul appears to be the best tool for coverage. It does however seem a bit fiddly to configure with Testem.

In the mean time, we could just continue to run your existing setup to output coverage.

I'd like to see travis stages set up, where coverage, linting, build, tests can all be different stages.

**Comment by asturur** [comment]
i ll check things one at time. coverage can still be run with the old setup, manually, is just for me to see.

i m packing ad travelling 10 days, if you do not hear from me  is normal.

**Comment by asturur** [comment]
image

for hanged you mean it stays like this but without going back to console prompt?

**Comment by asturur** [comment]
I made some mistake, as you can see from commits. But i think having lint available from the main list is a good thing.
I managed to skip installing most of the things for browser testing and linting, i would do more, but i m not sure having this kind of install is good:

npm install uglifyjs@x.x.x testem@x.x.x  for the browser testing would help so much in shortening test times.

**Comment by jedateach** [comment]
Thanks for sorting out the iText animation test issues.

For this pull request, I think we should keep the scope on Testem, and not linting. Travis has a new 'stages' feature which we should use to completely separate out the linting.

I'll rework your changes so that it excludes linting.

**Comment by jedateach** [comment]
I've preserved your linting changes here:
https://github.com/jedateach/fabric.js/commit/a9d61b27bfa7ab5bc0869e10d13c8f1db4e60982
However - I don't think this is how we should separate linting, but should instead focus on using Travis's stages feature:
https://docs.travis-ci.com/user/build-stages

**Comment by asturur** [comment]
i see there are build stages and we can keep also the installed deps from one stage to another

**Comment by asturur** [comment]
so i think we can merge this.
Coverage is not a test, nor something it has to be run from travis now.

Is something i look at when changing functions in order to keep it constant or grow.
If i found a way to make instabul return error if coverage do not meet a minimum percentage, i can make it a step.

Do you think we can merge or do you want to add something else?

**Comment by asturur** [comment]
i merge it, it looks good.
Thank you very much this change is much appreciated.
For now node-qunit and instabul are staying for manual running, then i ll try to remove them

**Comment by jedateach** [comment]
Cool 🎉

**Pull request merged**